### PR TITLE
Swarm gateways db fixes

### PR DESCRIPTION
**Pull request by zelig**

### CODE REVIEW

File: cmd/swarm/main.go [hunk 1]
```diff
@@ -248,6 +248,15 @@ Removes a path from the manifest
 Cleans database of corrupted entries.
 `,
 		},
+		{
+			Action:    importdb,
+			Name:      "importdb",
+			Usage:     "Imports chunks from one datastore to another",
+			ArgsUsage: "<source-datadir> s",
```
Comment (homotopycolimit) [inline cmd/swarm/main.go]: no more <source-dir> <target-dir>? 
What's the target? How do I import from A to B? Would I use importdb <source-dir> --datadir=/path/to/target ?

File: swarm/storage/dbstore.go [hunk 2]
```diff
@@ -513,3 +541,37 @@ func (self *dbSyncIterator) Next() (key Key) {
 	self.it.Release()
 	return nil
 }
+
+func Import(sourcepath string, targetpath string) (uint64, error) {
+	sourcedb, err := NewDbStore(sourcepath, MakeHashFunc(defaultHash), defaultDbCapacity, defaultRadius)
+	if err != nil {
+		return 0, err
+	}
+	targetdb, err := NewDbStore(targetpath, MakeHashFunc(defaultHash), defaultDbCapacity, defaultRadius)
```
Comment (cobordism) [inline swarm/storage/dbstore.go]: is defaultDbCapacity here equal to our default of 20GB? because I'd want my targetDB to be bigger than that ( I set it to be bigger in the config.json, but ... )

File: cmd/swarm/import.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+	sourcedir := args[0]
+	targetdir := args[1]
+
+	chunkcount, err := storage.Import(sourcedir, targetdir)
+	if err != nil {
+		utils.Fatalf("import failed: %s", err)
+	}
+
+	log.Trace(fmt.Sprintf("%d chunks imported", chunkcount))
```
Comment (lmars) [inline cmd/swarm/import.go]: This should be `log.Info` so you don't need to set `--verbosity` in order to see it.

File: cmd/swarm/main.go (hunk 1 above)
Comment (lmars) [inline cmd/swarm/main.go]: This `ArgsUsage` looks incomplete, should it be `<source-datadir> <target-datadir>`?

File: swarm/storage/dbstore.go [hunk 4]
```diff
@@ -320,44 +325,66 @@ func (s *DbStore) Put(chunk *Chunk) {
...
+	for range s.batchesC {
+		s.lock.RLock()
+		b := s.batch
+		e := s.entryCnt
+		d := s.dataIdx
+		a := s.accessCnt
+		c := s.batchC
+		s.batchC = make(chan bool)
+		s.batch = new(leveldb.Batch)
+		s.lock.RUnlock()
```
Comment (lmars) [inline swarm/storage/dbstore.go]: This code is assigning to fields on `DbStore` so shouldn't it hold the write lock?

File: swarm/storage/dbstore.go [hunk 5]
```diff
@@ -320,44 +325,66 @@ func (s *DbStore) Put(chunk *Chunk) {
...
+	log.Trace(fmt.Sprintf("DbStore: quit batch write loop"))
+}
+
+// must be called non concurrently
+func (s *DbStore) writeBatch(b *leveldb.Batch, entryCnt, dataIdx, accessCnt uint64) {
+	b.Put(keyEntryCnt, U64ToBytes(entryCnt))
+	b.Put(keyDataIdx, U64ToBytes(dataIdx))
+	b.Put(keyAccessCnt, U64ToBytes(accessCnt))
+	l := s.batch.Len()
+	_ = s.db.Write(b)
```
Comment (lmars) [inline swarm/storage/dbstore.go]: Does this call return an error? Why are we ignoring it?

File: swarm/storage/dbstore.go [hunk 6]
```diff
@@ -451,6 +478,7 @@ func (s *DbStore) getEntryCnt() uint64 {
 }
 
 func (s *DbStore) Close() {
+	close(s.batchesC)
```
Comment (lmars) [inline swarm/storage/dbstore.go]: It looks like this will race with `doPut` which tries to write to this channel.

File: swarm/storage/dbstore_test.go [hunk 7]
```diff
@@ -24,6 +24,11 @@ import (
 	"github.com/ethereum/go-ethereum/common"
 )
 
+//
+// func init() {
+// 	log.Root().SetHandler(log.LvlFilterHandler(log.LvlTrace, log.StreamHandler(os.Stderr, log.TerminalFormat(false))))
+// }
```
Comment (lmars) [inline swarm/storage/dbstore_test.go]: Why is this commented?

File: swarm/storage/dbstore_test.go [hunk 8]
```diff
@@ -58,9 +63,10 @@ func TestDbStore128_100_(t *testing.T) {
 	testDbStore(100, 128, t)
 }
 
-func TestDbStore2_100_(t *testing.T) {
-	testDbStore(100, 2, t)
-}
+//
+// func TestDbStore2_100_(t *testing.T) {
+// 	testDbStore(100, 2, t)
+// }
```
Comment (lmars) [inline swarm/storage/dbstore_test.go]: Why is this commented?

File: swarm/storage/dpa_test.go [hunk 9]
```diff
@@ -45,12 +46,15 @@ func TestDPArandom(t *testing.T) {
...
 
 	reader, slice := testDataReaderAndSlice(testDataSize)
+
 	wg := &sync.WaitGroup{}
 	key, err := dpa.Store(reader, testDataSize, wg, nil)
 	if err != nil {
 		t.Errorf("Store error: %v", err)
 	}
 	wg.Wait()
+	time.Sleep(2 * time.Second)
```
Comment (lmars) [inline swarm/storage/dpa_test.go]: Why is this necessary?

**Comment by lmars** [comment]
@zelig is there a reason that commit 3a678a1 is included in this pull request?

File: cmd/swarm/import.go (hunk 3 above)
Comment (zelig) [reply cmd/swarm/import.go]: i rebased on @lmars db tools PR

File: swarm/api/http/server.go [hunk 10]
```diff
@@ -522,6 +522,12 @@ func (s *Server) HandleGetList(w http.ResponseWriter, r *Request) {
 // HandleGetFile handles a GET request to bzz://<manifest>/<path> and responds
 // with the content of the file at <path> from the given <manifest>
 func (s *Server) HandleGetFile(w http.ResponseWriter, r *Request) {
+	// ensure the root path has a trailing slash so that relative URLs work
+	if r.uri.Path == "" && !strings.HasSuffix(r.URL.Path, "/") {
+		http.Redirect(w, &r.Request, r.URL.Path+"/", http.StatusMovedPermanently)
+		return
+	}
+
```
Comment (cobordism) [inline swarm/api/http/server.go]: Does this mean that a / is added 
- to every URL? 
- only to the first part of the URL (the 'domain')
- to every URL that references a manifest but not to URLs referencing other content.

I think we are doing the second, but we want the third option, yes?
I want mysite.eth --> mysite.eth/
I want mysite.eth/www --> mysite.eth/www/
I want mysite.eth/www/myimage.png to stay unchanged.

File: swarm/storage/dbstore.go (hunk 6 above)
Comment (cobordism) [reply swarm/storage/dbstore.go]: is there a quick way to confirm / fix this, lmars?

Comment (lmars) [reply swarm/storage/dbstore.go]: To confirm this, either run the tests with the race detector on, or just eyeball the code :)

To fix it, you usually have two channels: one gets closed here, and the writing goroutine selects on it and then stops writing to the other channel.
